Use generated PDFium license for NuGet packages

sungaila opened the pull request:
I was made aware of the generated PDFium LICENSE in #194. Right now the NuGet packages are licensed under Apache-2.0, but this is not correct as it excludes third-party libraries and the preamble of "The PDFium Authors".

These are the changes:
- `build-all.yml` now copies `LICENSE` and renames it to `LICENSE.txt`.
  - Note: The nuspec demands a .txt or .md file for a custom license.
- `LICENSE.txt` is copied into the NuGet package root directory.
- Set a custom license instead of `Apache-2.0`.
- Set `requireLicenseAcceptance` to `true`. The user must agree to the license to install the NuGet package.
- Replace `Benoit Blanchon` with `Benoît Blanchon` (we are not limited by ASCII).

Here is a screenshot of how this would look in Visual Studio 2022 (language is set to German). You are promted to agree to the custom license and can click on "Lizenz anzeigen" (show license) to read the generated PDFium license.

![image](https://github.com/user-attachments/assets/2c6aaed6-3a69-453a-b08d-eccd360d674c)

sungaila committed:
Use custom license file for NuGet packages

sungaila committed:
Rename LICENSE to LICENSE.txt with the bash script

sungaila committed:
Correct author name

sungaila committed:
Remove package

sungaila commented on .github/workflows/build-all.yml:
```diff
@@ -597,6 +597,9 @@ jobs:
         run: mkdir -p nuget/include/pdfium && cp -r nuget/pdfium-win-x86/include/* nuget/include/pdfium/
+      - name: Copy license file
+        shell: bash
+        run: cp nuget/pdfium-win-x86/LICENSE nuget/include/LICENSE.txt
```
I was not aware of that. The same header files are used for all build artifacts so I figured it would work for the license, too.

In that case maybe someone else could update the workflows and bash scripts.

bblanchon replied:
I think it would be easier if licenses were in multiple files, so you would just have to merge the folders to get the licenses for all the targets.
Do you guys think it would make sense to have a `licenses/` folder instead of a `LICENSE` file?

sungaila replied:
@bblanchon For the NuGet packages you must decide on a single license file, there is no option to have multiple files listed.

If possible, there should be a script that concats all licenses (split with something like `# BEGIN android-arm64`) so the NuGet specification is satisfied.

bblanchon replied:
The NuGet packages could point to the top level LICENSE, which would include a text like

> This package also includes third-party software. See the `licenses/` directory for their respective licenses.

Then we could bundle the `licenses/` folder in the NuGet package:

```xml
<files>
    <file src="licenses\**" target="licenses\" />
</files>
```

This will probably make life easier for the NuGet package users as well.

mara004 replied:
Possibly, yes. A single file is usually a bit easier to handle than a directory, but some licenses (if read strictly) might even need to be distributed as the exact file (see https://github.com/bblanchon/pdfium-binaries/issues/41#issuecomment-1153613908).

@sungaila Can you explain what speaks against splitting in architecture-specific packages?

sungaila replied:
@bblanchon Yes, I like that idea. The user would have to unzip the package to check for themselves but that should be fine.

@mara004 The convention is to split NuGet packages in platforms (e.g. Linux/Android/WebAssembly) instead of architectures (e.g. arm64/x64/musl-x64). The idea is that you if pick the macOS package then all expected archs should be included.

This is just convention though. However, there are about 7 million downloads right now (https://www.nuget.org/profiles/bblanchon) and introducing breaking changes should be avoided.

You could introduce new meta packages like `bblanchon.PDFium.macOS` pointing to `bblanchon.PDFium.macOS.x64` and `bblanchon.PDFium.macOS.arm64`. But this would be overkill imho.

bblanchon replied:
Grouping NuGet packages per platform is more convenient for everyone.
This license file issue doesn't justify splitting the packages.

bblanchon replied:
PS: an alternative would be to split the NuGet packages per architecture and add "meta" packages that group them per platform.

sungaila replied:
>(FWIW, I always try to avoid shipping unnecessary binaries that take disk space but aren't compatible with the host anyway.)

Good point, but in .NET you would want to ship all available native binaries for portability. Only when you publish for a specific platform (e.g. `linux-musl-x64`; doesn't matter if JIT or AOT compiled) the binaries are discarded.

I'm not disagreeing with you, I just want to point out Microsoft's design decisions. ;-)

mara004 replied:
> PS: an alternative would be to split the NuGet packages per architecture and add "meta" packages that group them per platform.

Yes, that's basically what @sungaila said in https://github.com/bblanchon/pdfium-binaries/pull/202#discussion_r2075661478, right? 😅

> You could introduce new meta packages like bblanchon.PDFium.macOS pointing to bblanchon.PDFium.macOS.x64 and bblanchon.PDFium.macOS.arm64. But this would be overkill imho.

sungaila replied:
>PS: an alternative would be to split the NuGet packages per architecture and add "meta" packages that group them per platform.

@bblanchon I could do that (would need some time though). On the positive side there wouldn't be any changes needed for the workflow/script side of things.

Also it wouldn't be a breaking change for users updating their NuGet packages.

bblanchon replied:
Let's focus back on the license issue.
I get a feeling you guys agreed that a `licenses/` folder would be a good step forward, right?

sungaila replied:
@bblanchon Yes, I'd rather have a licenses folder than ~20 new packages. That's a fair compromise, I think.

bblanchon replied:
OK, I'll split the license file then.
Just give me a few days, as I'm pretty busy at the moment.
This will also be the opportunity to solve #194.

bblanchon commented:
I replaced the aggregated `LICENSE` with a `licenses/` folder in the tarballs.
We should see the result in tomorrow's build.

sungaila committed:
Merge with master

sungaila committed:
Include licenses folder in NuGet packages

sungaila commented:
> I replaced the aggregated `LICENSE` with a `licenses/` folder in the tarballs. We should see the result in tomorrow's build.

Looks fine to me, good job! I've updated the nuspec files to include the licenses directory. The resulting directories look like this:

- bblanchon.PDFium.Win32.1.0.0.nupkg\licenses\pdfium-win-arm64\licenses\
- bblanchon.PDFium.Win32.1.0.0.nupkg\licenses\pdfium-win-x64\licenses\
- bblanchon.PDFium.Win32.1.0.0.nupkg\licenses\pdfium-win-x86\licenses\

It doesn't look as nice as I would have liked, but it deals just fine with cases where the licenses differ between architectures (arm64, x64, x86).

sungaila commented:
> I was under the impression the idea was to merge the licenses folders? For otherwise, you could as well have shipped multiple aggregated license files along with the builds...

Merging them wouldn't make sense since the licenses are different between archs.

I don't care if a single file or multiple files as long you can match the licenses to each binary.

bblanchon commented:
> I think it would be easier if licenses were in multiple files, so you would just have to merge the folders to get the licenses for all the targets.

This was also my initial idea: provide the union of the licenses for all the architectures supported by one package.

> Merging them wouldn't make sense since the licenses are different between archs.

They vary significantly between OSes and v8/non-v8, but not so much between archs.
In fact, they are all the same for the example you gave; and it's probably the same for the other OSes.

Even if they differed from one arch to the other, it would be only one or two libraries, so it's not worth the trouble.
